### PR TITLE
VideoCommon: Apply custom texture scale for arbitrary mipmaps.

### PR DESCRIPTION
**Pull request by degasus**
We want to get the same mipmap level. And if the IR and the custom
texture upscaling fits, we don't need to modify the LOD bias.

### CODE REVIEW

**Comment by degasus** [comment]
@Tomcc and @BigheadSMZ

**Comment by BigheadSMZ** [comment]
I tested this but I don't know what I'm supposed to be looking for. Is it supposed to get the same effect as the mod/screenshots [here](https://forums.dolphin-emu.org/Thread-dolphin-custom-texture-mipmaps?pid=460600#pid460600)? If so, no luck. IR/AF is still breaking (as in pushing/pulling) the effects in the same way [as it always did](https://i.imgur.com/iBFDknN.jpg). 

If I understand correctly, the idea of this PR is to find game textures that have arb mipmap effects and apply the arbitrary mipmap fix to custom textures that replace those as well? If so, I don't believe trying to automate it would be the best route. This would be great for SMG 1/2, but games that did not have mipmap effects that we wish to add them to, we still won't have control for those. Plus, some may not want the effects if the game had them. This is why I proposed the idea of some kind of flag in the texture name, so when to force mipmaps at a fixed distance can be selectively chosen on a per-texture basis.

But if this is not what this PR is trying to achieve, then ignore this comment. :P

**Comment by degasus** [comment]
Yeah, this is for the SMG 1/2 case. On master, we will pick the custom mipmap with the same size for arbitrary textures. With this PR, we will pick the same mipmap level.

File: Source/Core/VideoCommon/VertexManagerBase.cpp [hunk 1]
```diff
@@ -257,8 +258,10 @@ static void SetSamplerState(u32 index, bool custom_tex, bool has_arbitrary_mips)
     // Apply a secondary bias calculated from the IR scale to pull inwards mipmaps
     // that have arbitrary contents, eg. are used for fog effects where the
     // distance they kick in at is important to preserve at any resolution.
+    // Correct this with the upscaling factor of custom textures.
     state.lod_bias =
-        state.lod_bias + std::log2(static_cast<float>(g_renderer->GetEFBScale())) * 256.f;
+        state.lod_bias +
+        std::log2(static_cast<float>(g_renderer->GetEFBScale()) / custom_tex_scale) * 256.f;
```
Comment (degasus) [inline Source/Core/VideoCommon/VertexManagerBase.cpp]: This is fixed-point logic. Please clamp.

**Comment by JosJuice** [comment]
TextureCacheBase.cpp's ArbitraryMipmapDetector never gets any levels added into it when a custom texture is used (in other words, custom textures are treated as if they never have arbitrary mipmaps), which explains why Bighead's testing didn't show any differences between this PR and master. So in addition to this fix, we need a PR that enables arbitrary mipmap detection for custom textures, either by using ArbitraryMipmapDetector, or by checking file names like Bighead suggested.

**Comment by BigheadSMZ** [comment]
I still wasn't sure exactly what this did, so I merged these changes with the latest master to test it out since the auto-builds didn't load "_arb" textures. I gotta say this is awesome. It completely eliminates the guesswork involved when trying to figure out which mipmap levels should have which images, which is especially useful for the SMG games. I have to rework the textures I've already done, but it's totally worth it since this allows higher resolution mipmaps to stretch farther. It also works well for textures that did not originally have arbitrary mipmaps.

**Comment by degasus** [comment]
Rebased and added the clamp.

@BigheadSMZ May you test it again?

**Comment by BigheadSMZ** [comment]
Tested about 20 different textures in different scenarios between SMG1/2 and Xenoblade and can't find any issues. Works perfect as far as I can tell. :smiley: :thumbsup:

**Pull request merged**